homeassistant/components/ibeacon/coordinator.py

```diff
@@ -0,0 +1,378 @@
+
+
+@callback
+def _async_dispatch_update(
+    hass: HomeAssistant,
+    device_id: str,
+    service_info: bluetooth.BluetoothServiceInfoBleak,
+    parsed: iBeaconAdvertisement,
+    new: bool,
+    unique_address: bool,
+) -> None:
+    """Dispatch an update."""
+    if new:
+        async_dispatcher_send(
+            hass,
+            SIGNAL_IBEACON_DEVICE_NEW,
+            device_id,
+            async_name(service_info, parsed, unique_address),
+            parsed,
+        )
+        return
+
+    async_dispatcher_send(
+        hass,
+        signal_seen(device_id),
+        parsed,
+    )
+
+
+class IBeaconCoordinator:
+    """Set up the iBeacon Coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, entry: ConfigEntry, registry: DeviceRegistry
+    ) -> None:
+        """Initialize the Coordinator."""
+        self.hass = hass
+        self._entry = entry
+        self._min_rssi = entry.options.get(CONF_MIN_RSSI) or DEFAULT_MIN_RSSI
+        self._dev_reg = registry
+
+        # iBeacon devices that do not follow the spec
+        # and broadcast custom data in the major and minor fields
+        self._ignore_addresses: set[str] = set(
+            entry.data.get(CONF_IGNORE_ADDRESSES, [])
+        )
+
+        # iBeacons with fixed MAC addresses
+        self._last_rssi_by_unique_id: dict[str, int] = {}
+        self._group_ids_by_address: dict[str, set[str]] = {}
+        self._unique_ids_by_address: dict[str, set[str]] = {}
+        self._unique_ids_by_group_id: dict[str, set[str]] = {}
+        self._addresses_by_group_id: dict[str, set[str]] = {}
+        self._unavailable_trackers: dict[str, CALLBACK_TYPE] = {}
+
+        # iBeacon with random MAC addresses
+        self._group_ids_random_macs: set[str] = set()
+        self._last_seen_by_group_id: dict[str, bluetooth.BluetoothServiceInfoBleak] = {}
+        self._unavailable_group_ids: set[str] = set()
+
+    @callback
+    def _async_handle_unavailable(
+        self, service_info: bluetooth.BluetoothServiceInfoBleak
+    ) -> None:
+        """Handle unavailable devices."""
+        address = service_info.address
+        self._async_cancel_unavailable_tracker(address)
+        for unique_id in self._unique_ids_by_address[address]:
+            async_dispatcher_send(self.hass, signal_unavailable(unique_id))
+
+    @callback
+    def _async_cancel_unavailable_tracker(self, address: str) -> None:
+        """Cancel unavailable tracking for an address."""
+        self._unavailable_trackers.pop(address)()
+
+    @callback
+    def _async_ignore_address(self, address: str) -> None:
+        """Ignore an address that does not follow the spec and any entities created by it."""
+        self._ignore_addresses.add(address)
+        self._async_cancel_unavailable_tracker(address)
+        self.hass.config_entries.async_update_entry(
+            self._entry,
+            data=self._entry.data
+            | {CONF_IGNORE_ADDRESSES: sorted(self._ignore_addresses)},
```

Comment:
Splitting a dict update without parenthesis inside a call on more than one line is hard to read, as the indentation is missing.